cli/test_runner.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+
+use crate::fs as deno_fs;
+use crate::installer::is_remote_url;
+use globset;
+use std;
+use std::path::PathBuf;
+use url::Url;
+use walkdir::WalkDir;
+
+fn find_local_test_modules(globs: Vec<String>) -> Vec<Url> {
+  use globset::{Glob, GlobSetBuilder};
+  dbg!(globs.clone());
+  let mut builder = GlobSetBuilder::new();
+  // A GlobBuilder can be used to configure each glob's match semantics
+  // independently.
+  let cwd = std::env::current_dir().unwrap();
+
+  for glob_string in globs {
+    let glob_path = cwd.join(glob_string);
+    let g = Glob::new(&glob_path.to_string_lossy()).expect("Bad glob string");
+    builder.add(g.clone());
+    dbg!(g.glob(), g.regex());
+  }
+  let glob_set = builder.build().expect("Failed to build glob");
+
+  WalkDir::new(".")
+    .into_iter()
+    .filter_map(|v| v.ok())
+    .map(|e| cwd.join(e.path()).canonicalize())
+    .filter_map(|v| v.ok())
+    .filter(|p| {
+      let result = glob_set.is_match(p);
+      dbg!(p, result.clone());
+      result
+    })
+    .map(|p| {
+      dbg!("mapping");
+      Url::from_file_path(p).unwrap()
+    })
+    .collect()
```

Comment:
Isn't this https://github.com/denoland/deno_std/issues/576 again? What's wrong with the `glob` crate used for the formatter?

Reply:
```
Comparison with the glob crate
Supports alternate "or" globs, e.g., *.{foo,bar}.
Can match non-UTF-8 file paths correctly.
Supports matching multiple globs at once.
Doesn't provide a recursive directory iterator of matching file paths, although I believe this crate should grow one eventually.
Supports case insensitive and require-literal-separator match options, but doesn't support the require-literal-leading-dot option.
```
Our default test glob is: `**/?(*_)test.{js,ts}`. `glob` in `fmt` will be replaced with `globset`.

Reply:
Honestly that glob I wrote uses slightly inconsistent syntax. The proper [extended globbing](https://www.linuxjournal.com/content/bash-extended-globbing) syntax for OR is `@(a|b)`.

Does the `glob` crate support extended globbing? In other words would it support
`**/?(*_)test.@(js|ts)`

Reply:
> Honestly that glob I wrote uses slightly inconsistent syntax. The proper [extended globbing](https://www.linuxjournal.com/content/bash-extended-globbing) syntax for OR is `@(a|b)`.
> 
> Does the `glob` crate support extended globbing? In other words would it support
> `**/?(*_)test.@(js|ts)`

I don't think so. I'm trying to port behavior of current test runner but `glob` in `std/` is whole other beast than the ones available in Rust - it requires a log of plumbing to be written

Reply:
Hmm, I see it doesn't. It doesn't to do much at all...

Reply:
Yeah... `globset` is not really easier either; I can't really get it to work
```
[cli/test_runner.rs:30] g.glob() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/**/?(*_)test.{js,ts}"
[cli/test_runner.rs:30] g.regex() = "(?-u)^/Users/biwanczuk/dev/deno/cli/tests/test_runner(?:/|/.*/).\\(.*_\\)test\\.(ts|js)$"
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/foo.ts"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/test.js"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/bar_test.js"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/subdir"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/subdir/foo.ts"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/subdir/test.js"
[cli/test_runner.rs:39] result.clone() = false
[cli/test_runner.rs:39] p.path() = "/Users/biwanczuk/dev/deno/cli/tests/test_runner/subdir/bar_test.js"
```

Reply:
With the `globset`/walking method, you can just costlessly expand the glob into `["**/test.js", "**/test.ts", "**/*_test.js", "**/*_test.ts"]` like it used to be.

Reply:
Right, now it works :))